ff/src/fields/sqrt.rs

```diff
@@ -84,66 +125,230 @@ impl<F: crate::Field> SqrtPrecomputation<F> {
+            // which means that no square root exists.
+            return None;
+        }
+        let j = v - k;
+        w = z;
+        for _ in 1..j {
+            w.square_in_place();
+        }
+
+        z = w.square();
+        b *= &z;
+        x *= &w;
+        v = k;
+    }
+    // Is x the square root? If so, return it.
+    if x.square() == *elem {
+        Some(x)
+    } else {
+        // Consistency check that if no square root is found,
+        // it is because none exists.
+        debug_assert!(!matches!(elem.legendre(), LegendreSymbol::QuadraticResidue));
+        None
+    }
+}
+
+fn shanks<F: crate::Field>(
+    elem: &F,
+    char_minus_three_div_four: &[u64],
+    deg_minus_three_div_two_plus_one: usize,
+) -> Option<F> {
+    // Computing a1 = Using decomposition of (q-3)/4 = a + p[pa + (3a+2)] * sum_i=1^(m-3)/2 p^2i
+    // where a = (p - 3) / 4.
+    // factor1 = elem^a
+    let factor1 = elem.pow(char_minus_three_div_four);
+    // elem_to_p = elem^p
+    let elem_to_p = elem.frobenius_map(1);
+    // factor2_base = elem^(p^2)a * elem^3pa * elem^2p
+    let factor2_base = elem_to_p.frobenius_map(1).pow(char_minus_three_div_four)
+        * elem_to_p.pow(&[3u64]).pow(char_minus_three_div_four)
+        * elem_to_p.square();
+    // factor2 = prod_i=1^(m-3)/2 factor2_base^(p^2i)
+    let mut factor2 = F::one();
+    for i in 1..deg_minus_three_div_two_plus_one {
+        factor2 *= factor2_base.frobenius_map(i * 2 as usize);
+    }
+    let a1 = factor1 * factor2;
+
+    let a1_elem = a1 * elem;
+    let a0 = a1 * a1_elem;
+    if a0 == -F::one() {
+        return None;
+    }
+
+    Some(a1_elem)
+}
+
+fn atkin<F: crate::Field>(
+    elem: &F,
+    trace: &F,
+    char_minus_five_div_eight: &[u64],
+    deg_minus_three_div_two_plus_one: usize,
+) -> Option<F> {
+    // Computing a1 = elem^(q-5)/8 using decomposition of
+    // (q-5)/8 = a + p[pa + (5a+3)] * sum_i=1^(m-3)/2 p^2i
+    // where a = (p - 5) / 8.
+    // factor1 = elem^a
+    let factor1 = elem.pow(char_minus_five_div_eight);
+    // elem_to_p = elem^p
+    let elem_to_p = elem.frobenius_map(1);
+    // factor2_base = elem^(p^2)a * elem^5pa * elem^3p
+    let factor2_base = elem_to_p.frobenius_map(1).pow(char_minus_five_div_eight)
+        * elem_to_p.pow(&[5u64]).pow(char_minus_five_div_eight)
+        * elem_to_p.pow(&[3u64]);
+    // factor2 = prod_i=1^(m-3)/2 factor2_base^(p^2i)
+    let mut factor2 = F::one();
+    for i in 1..deg_minus_three_div_two_plus_one {
+        factor2 *= factor2_base.frobenius_map(2 * i);
+    }
+    let a1 = factor1 * factor2;
+
+    let a0 = (a1.square() * elem).square();
+    if a0 == -F::one() {
+        return None;
+    }
+
+    let b = a1 * trace;
+    let elem_b = b * elem;
+    let i = elem_b.double() * b;
+    let x = elem_b * (i - F::one());
+
+    Some(x)
+}
+
+fn kong<F: crate::Field>(
+    elem: &F,
+    trace: &F,
+    c: &F,
+    d: &F,
```

Comment:
I considered expanding out `d` into a description of how it's related to `c` but `c_to_order_minus_nine_div_eight` seems too long

Reply:
We can just clarify that in a comment in `d`.

Reply:
I already did